### DISCUSSION TITLE
RFC: Pie on mobile — running the runtime and a Metal-class driver on iOS and Android

### DISCUSSION THREAD

**Discussion by aarushkandukoori** [pie-project/pie#526] [Ideas]
**Author:** Aarush Kandukoori (CMU undergrad, CS)
**Status:** Proposal — seeking maintainer feedback before building further
**Scope:** iOS first (aligned with the existing Metal investment), Android second
**Patch branch:** [aarushkandukoori/pie@ios-pulley](https://github.com/aarushkandukoori/pie/tree/ios-pulley)

## Summary

I propose porting Pie to phones: the wasmtime-based runtime running inferlets on-device, backed by a mobile GPU driver. The runtime is closer to this than it looks — I verified that the `pie` runtime crate **cross-compiles for `aarch64-apple-ios` today with a ~10-line patch** (diff below), including wasmtime configured for the Pulley interpreter, which is what makes inferlet execution legal under iOS's no-JIT rules.

The end state I want to build toward is the roadmap's own line — *"durable inferlets that migrate freely across runtimes"* — demonstrated across the device boundary: an inferlet starts on a Mac (or datacenter) Pie node, persists its KV state, and resumes on an iPhone, offline.

## Motivation

1. **On-device serving is a real gap.** Every mobile LLM stack today (llama.cpp apps, MLC-LLM, LiteRT-LM, Apple Foundation Models) is a fixed chat pipeline. None of them let an application program the serving loop — branch/rewind KV, custom sampling, agent loops with tool calls — which is exactly Pie's thesis. A phone is arguably where programmable serving matters *most*: privacy-sensitive data (health sensors, messages, location) that shouldn't leave the device, and connectivity that comes and goes.
2. **Pie's architecture is unusually well-suited to iOS.** The commonly assumed blocker — "you can't run arbitrary user code on iOS" — mostly dissolves here, because Pie's user code is control-plane. The FLOPs live in the native driver; inferlets orchestrate. An interpreter-executed inferlet is acceptable in a way an interpreter-executed matmul never would be (and the 0.4 speculative-execution work that removed ~90% of programmability overhead insulates this further).
3. **It extends 0.5's direction rather than competing with it.** The 0.5 distributed design (controller, gateway, worker-to-worker KV transport) gives phones a natural role: a phone is just a small worker with an unreliable link. Durable-inferlet migration across the device boundary is the same mechanism, harder network.

### Runtime: wasmtime with per-OS execution backends

iOS forbids writable+executable pages in distributable apps, so Cranelift JIT and loading precompiled `.cwasm` are both out. wasmtime's **Pulley** interpreter (Tier 2) executes wasm as data — no executable mappings — and is the established pattern for wasmtime-on-iOS. Android has no such restriction; Cranelift runs unchanged at full speed.

Verified so far (Pie @ `cdf7ca3`, Rust 1.96, iOS SDK 26.0):

- `cargo check -p pie --target aarch64-apple-ios` — **passes with zero code changes.** The whole host dep tree (wasmtime 41.0.3, tokio, `pie-bridge`'s rkyv/shmem code, rustls, ring) compiles for iOS out of the box.
- With `pulley` added to the wasmtime features and the engine targeting `pulley64` on iOS — **also passes.** Full diff:

```diff
--- a/runtime/Cargo.toml
+++ b/runtime/Cargo.toml
-wasmtime = { version = "41.0.3", features = ["pooling-allocator"] }
+wasmtime = { version = "41.0.3", features = ["pooling-allocator", "pulley"] }

--- a/runtime/src/bootstrap.rs   (init_wasmtime)
+    // iOS forbids writable+executable pages (no JIT), so execute inferlets
+    // with wasmtime's Pulley interpreter: wasm runs as data, which is
+    // App-Store-compatible. Inferlets are control-plane code — the FLOPs
+    // live in the driver — so the interpreter tax stays off the hot path.
+    #[cfg(target_os = "ios")]
+    wasm_config
+        .target("pulley64")
+        .expect("wasmtime built without the `pulley` feature");
```

Known costs and open items on the runtime side:

- Pulley is roughly an order of magnitude slower than Cranelift and doesn't support the threads proposal. Acceptable for control-plane inferlets; worth measuring per-token overhead explicitly (M4 below).
- The pooling allocator's virtual-memory reservations and `wasm_max_memory_mb` defaults need iOS-appropriate tuning (jetsam kills on physical footprint; defaults sized for servers).
- Both mobile targets are wasmtime Tier 3, so this work should carry its own device CI.
- Embedding shape: `pie` builds as an rlib; the iOS app would embed it as a static library behind a small Swift shim. The HTTP/WebSocket serving surface can stay for LAN use (a phone serving inferlets to the local network is a fun demo on its own) but isn't required for on-device use.

### Driver: two tracks

**Track A — bring-up driver on llama.cpp (fastest path to end-to-end).**
Pie's driver seam is small and explicit: a C entry pair (`pie_driver_<flavor>_run_inproc` / `_request_stop`) speaking the `pie_bridge` rkyv schema, registered in `server/src/driver_ffi.rs` (`Flavor`), `server/src/config.rs` (`DriverKind`), and `server/src/embedded_driver.rs`, with `driver/dummy/src/lib.rs` as a pure-Rust template. llama.cpp is the only mobile-mature backend whose public C API already exposes Pie's KV primitives:

| Pie concept | llama.cpp API |
|---|---|
| branch | `llama_memory_seq_cp` |
| rewind | `llama_memory_seq_rm` / `llama_memory_seq_add` |
| persist / reopen | `llama_state_seq_get_data` / `set_data` / `save_file` / `load_file` |
| custom sampling | raw logits + sampler-chain API |
| batched branches | multi-sequence `llama_batch` |

Granularity is sequence-level rather than page-level, so Pie's page abstraction sits above it — fine for bring-up, not the end state. Backends: Metal on iOS; Vulkan plus Qualcomm's OpenCL-Adreno path on Android; CPU fallback everywhere.

**Track B — port `driver/metal` to iOS (the real target).**
Since 0.5 consolidates on cuda + metal, the durable investment is making the native Metal driver run on iOS. Most of the delta looks like build-system work plus a serious memory-budget model — which overlaps with the already-acknowledged "host fit check is overly conservative" gap, except on iOS the budget is adversarial (jetsam) rather than merely conservative. One question I'd like maintainer input on: which Metal 4 features does `driver/metal` require, and what's the implied iPhone/iPad hardware floor?

### Distributed fit

Under the 0.5 controller/gateway model, a phone registers as a worker with a capability manifest (small quantized models, one concurrent inferlet, unreliable link). Worker-to-worker KV transport falls back from NIXL/RDMA to TCP/QUIC. Durable-inferlet migration to/from the phone is then the roadmap mechanism with a harder network — and the flagship demo (M3).

### Platform constraints, stated honestly

- **App Store 2.5.2:** a bundled interpreter executing bundled wasm is the settled pattern (a-Shell ships wasm interpreters on the App Store; Wasmer supports iOS). Downloading inferlets at runtime is the gray zone — v1 bundles inferlets; research builds via Xcode/TestFlight don't care.
- **Memory:** jetsam limits mean 1–4B-class models at Q4 (Qwen3, Gemma 3n territory), and the fit check must be a hard gate, not advisory.
- **Backgrounding/thermals:** iOS will suspend or throttle long generations; sustained-load behavior is part of the benchmark plan, not an afterthought.

## Milestones

- **M0 — runtime cross-compiles for iOS (done, diff above).**
- **M1 — first token on a phone:** headless iOS test app embedding the runtime + a Track-A llama.cpp driver; one existing inferlet runs unmodified (that's the wasm portability payoff).
- **M2 — `driver/metal` on iOS:** build-system port + iOS memory-budget model; retire Track A on iOS.
- **M3 — durable-inferlet migration demo:** persist on macOS Pie, transfer, reopen and continue on iPhone, offline.
- **M4 — reproducible perf/energy harness:** TTFT, prefill/decode tok/s, joules/token, thermal-sustain curves vs. llama.cpp and MLC-LLM baselines on identical models/quants; published as scripts, addressing the known benchmarks-reproducibility gap.

Android follows the same shape with easier constraints (Cranelift, no jetsam-class limits, Vulkan/OpenCL via Track A).

## Questions for maintainers

1. Given the 0.5 rewrite in flight, what should this target — `main` as it stands, or is there a branch/timeline where the new driver layout and PTIR model stabilize enough to build against?
2. Would a `driver/llamacpp` bring-up driver be acceptable upstream (perhaps as an explicitly-temporary flavor), or would you rather see iOS effort go straight into the Metal driver port?
3. Does the 0.5 controller's placement model have room for asymmetric, intermittently-connected workers, or should phone nodes live behind a separate gateway abstraction for now?
4. Is there prior internal thinking on mobile/edge targets I should read before going further?

## About me

CMU CS undergrad. Prior work: a patent-pending wearable medical device doing real-time QRS detection with a 1D-CNN beat classifier — which is also the use case I care about here: health-sensor agents whose data never leaves the device. I'm proposing to drive this as a sustained open-source contribution, starting with the smaller PRs (Metal fit-check budget model, Metal reference docs) so the mobile work rests on parts of the codebase I've already touched.

**Comment by aarushkandukoori** [comment]
**Progress update: real model inference now works on iOS (simulator).**

Qwen3-0.6B (Q4_K_M GGUF) generates through the full Pie stack inside an iPhone 16 Pro simulator app: the engine and WebSocket control plane boot in-process, `text-completion@0.2.15` executes under wasmtime's **Pulley** interpreter (no JIT, App Store compatible), and the **portable ggml driver cross-compiled for iOS** runs forward passes on CPU. Token deltas stream live into a SwiftUI view.

Numbers (release build, arm64 simulator on an M-series host): cold boot + 574 MB weight load + 96 tokens in **~8 s**; warm-engine 48-token generation in **~2 s** (roughly 15-20 tok/s decode).

Everything lives on [`aarushkandukoori/pie@ios-pulley`](https://github.com/aarushkandukoori/pie/tree/ios-pulley). What the port needed, all small:

- `pulley64` engine target on iOS (`runtime/src/bootstrap.rs`). Inferlets are control-plane code, so the interpreter cost stays off the hot path.
- File-backed mmap fallback for POSIX shmem (`driver/bridge/src/ipc/posix.rs`). The iOS sandbox denies `shm_open` with EPERM; embedded drivers are in-process, so a `$TMPDIR`-backed `MAP_SHARED` region is semantically equivalent.
- iOS cross-compile support for the portable driver's CMake build (`server/build.rs`: SDK sysroot defines, Accelerate and libc++ links; CPU-only ggml for now).
- `ios/pie-shim` (C-ABI staticlib embedding the engine, boot path mirrors `pie run`, streaming callback API) and `ios/demo-app` (minimal SwiftUI shell, no Xcode project needed).

![Qwen3 haiku through Pie on the iPhone simulator](https://raw.githubusercontent.com/aarushkandukoori/pie/ios-pulley/ios/demo-app/screenshots/haiku.png)

39-second demo video (cold boot, live token streaming, warm haiku): https://github.com/aarushkandukoori/pie/blob/ios-pulley/ios/demo-app/screenshots/demo.mp4

Next: the ggml Metal backend on iOS, physical devices, Android, then the durable-inferlet migration demo from the RFC. Question 1 above is now the one that matters most: which branch should this work target given the 0.5 driver reorg?

**Comment by ingim** [comment]
Hi @aarushkandukoori,

The demo looks fantastic! Thanks for the interesting proposal.

We've just rewritten the Pie for version 0.5, which will become our stable architecture. Would you be interested in leading iOS and Android support for Pie 0.5?

The new backend no longer uses ggml, so the scope of work may be broader than originally proposed.

P.S. Sorry for the late reply. I’ve just enabled notifications for new discussion threads.
